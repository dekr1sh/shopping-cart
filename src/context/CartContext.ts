import { createContext } from "react"
import type { Product } from "../data/products.ts"  // specifying the 'type' keyword optimizes the build

export interface CartItem extends Product {
  size: string       // shoe size
  quantity: number   // number of shoes
}

interface CartContextType {
  cartItems: CartItem[]
  addToCart: (product: Product, size: string, quantity: number) => void
  removeFromCart: (id: number, size: string) => void
  increaseQuantity: (id: number, size: string) => void
  decreaseQuantity: (id: number, size: string) => void
  clearCart: () => void
}

export const CartContext = createContext<CartContextType | undefined>(undefined)
