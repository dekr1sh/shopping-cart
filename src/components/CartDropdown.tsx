import { useCart } from "@/context/useCart";
import { Button } from "@/components/ui/button";
import { Minus, Plus, X, ShoppingBag } from "lucide-react";
import placeholder from "@/assets/other-images/placeholder.png"

interface CartDropdownProps {
  onClose: () => void;
}

export default function CartDropdown({ onClose }: CartDropdownProps) {
  const { cartItems, increaseQuantity, decreaseQuantity, removeFromCart, clearCart } = useCart();

  const totalPrice = cartItems.reduce((total, item) => total + item.price * item.quantity, 0);

  return (
    <div className="absolute right-0 mt-2 w-80 sm:w-96 bg-background border rounded-lg shadow-lg z-50">
      <div className="p-4 max-h-[80vh] overflow-auto">
        <div className="flex justify-between items-center mb-4">
          <h3 className="font-bold text-lg">Your Cart</h3>
          <Button variant="ghost" size="icon" onClick={onClose} aria-label="Close cart" className="cursor-pointer">
            <X size={18} />
          </Button>
        </div>

        {cartItems.length === 0 ? (
          <div className="py-8 text-center">
            <ShoppingBag className="mx-auto mb-4 text-muted-foreground" size={48} />
            <p className="text-muted-foreground">Your cart is empty</p>
            <Button className="mt-4 hover:bg-primary/85 cursor-pointer" onClick={onClose}>
              Continue Shopping
            </Button>
          </div>
        ) : (
          <>
            <div className="space-y-4 mb-4">
              {cartItems.map((item) => (
                <div key={`${item.id}-${item.size}`} className="flex items-center gap-4 border-b pb-4">
                  <div className="relative h-16 w-16 bg-muted rounded">
                    <img
                      src={item.image || placeholder}
                      alt={item.name}
                      className="object-cover rounded h-full w-full"
                    />
                  </div>
                  <div className="flex-1">
                    <h4 className="font-medium">{item.name}</h4>
                    <p className="text-sm text-muted-foreground">
                      {item.brand} • Size: {item.size}
                    </p>
                    <div className="flex justify-between items-center mt-2">
                      <div className="flex items-center border rounded">
                        <Button
                          variant="ghost"
                          size="icon"
                          className="h-8 w-8 rounded-none cursor-pointer"
                          onClick={() => decreaseQuantity(item.id, item.size)}
                          disabled={item.quantity <= 1}
                          aria-label="Decrease quantity"
                        >
                          <Minus size={14} />
                        </Button>
                        <span className="w-8 text-center">{item.quantity}</span>
                        <Button
                          variant="ghost"
                          size="icon"
                          className="h-8 w-8 rounded-none cursor-pointer"
                          onClick={() => increaseQuantity(item.id, item.size)}
                          aria-label="Increase quantity"
                        >
                          <Plus size={14} />
                        </Button>
                      </div>
                      <div className="text-right">
                        <p className="font-medium">₹{(item.price * item.quantity).toFixed(2)}</p>
                        <button
                          className="text-xs text-muted-foreground hover:text-destructive cursor-pointer"
                          onClick={() => removeFromCart(item.id, item.size)}
                        >
                          Remove
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              ))}
            </div>

            <div className="border-t pt-4">
              <div className="flex justify-between mb-2">
                <span>Subtotal</span>
                <span className="font-medium">₹{totalPrice.toFixed(2)}</span>
              </div>
              <div className="flex justify-between mb-4">
                <span>Shipping</span>
                <span>Calculated at checkout</span>
              </div>
              <div className="flex justify-between text-lg font-bold mb-4">
                <span>Total</span>
                <span>₹{totalPrice.toFixed(2)}</span>
              </div>
              <Button className="w-full mb-2 cursor-pointer">Checkout</Button>
              <div className="flex gap-2 mb-2">
                <Button variant="outline" className="flex-1 cursor-pointer" onClick={onClose}>
                  Continue Shopping
                </Button>
                <Button variant="destructive" className="flex-1 cursor-pointer" onClick={clearCart}>
                  Clear Cart
                </Button>
              </div>
            </div>
          </>
        )}
      </div>
    </div>
  );
}
