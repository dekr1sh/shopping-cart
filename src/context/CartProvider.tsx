import { useState, useEffect, type ReactNode } from "react"
import { CartContext, type CartItem } from "./CartContext.ts"
import type { Product } from "../data/products.ts"

export const CartProvider = ({ children }: { children: ReactNode }) => {
  const [cartItems, setCartItems] = useState<CartItem[]>([])
  const [totalItems, setTotalItems] = useState(0)
  const [totalPrice, setTotalPrice] = useState(0)
  const [isOpen, setIsOpen] = useState(false)

  useEffect(() => {
    const savedCart = localStorage.getItem("cart")
    if (savedCart) {
      setCartItems(JSON.parse(savedCart))
    }
  }, [])

  useEffect(() => {
    localStorage.setItem("cart", JSON.stringify(cartItems))

    const totItems = cartItems.reduce((total, item) => total + item.quantity, 0)
    setTotalItems(totItems)

    const price = cartItems.reduce((total, item) => total + item.price * item.quantity, 0)
    setTotalPrice(price)
  }, [cartItems])

  const addToCart = (product: Product, size: string, quantity: number) => {
    setCartItems((prevItems) => {
      const existingItemIndex = prevItems.findIndex((item) => item.id === product.id && item.size === size)

      if (existingItemIndex >= 0) {
        const updatedItems = [...prevItems]
        updatedItems[existingItemIndex].quantity += quantity
        return updatedItems
      } else {
        return [...prevItems, { ...product, size, quantity }]
      }
    })
  }

  const removeFromCart = (productId: string) => {
    setCartItems((prevItems) => prevItems.filter((item) => item.id !== +productId))
  }

  const updateQuantity = (productId: string, quantity: number) => {
    setCartItems((prevItems) =>
      prevItems.map((item) => (item.id === +productId ? { ...item, quantity: Math.max(1, quantity) } : item)),
    )
  }

  const clearCart = () => {
    setCartItems([])
  }

  return (
    <CartContext.Provider
      value={{ cartItems, totalItems, totalPrice, isOpen, setIsOpen, addToCart, removeFromCart, updateQuantity, clearCart }}
    >
      {children}
    </CartContext.Provider>
  )
}
