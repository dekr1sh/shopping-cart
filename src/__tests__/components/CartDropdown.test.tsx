import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { render } from "@/__tests__/test-utils";
import CartDropdown from "@/components/CartDropdown";
import { useCart } from "@/context/useCart";
import { describe, it, expect, vi, Mock } from "vitest";

// This mocks a module. Mocking a module means mocking every export from the module
vi.mock("@/context/useCart", () => ({ useCart: vi.fn() }));  

describe("CartDropdown", () => {
  it("renders correctly when cart is empty", () => {
    // `useCart as Mock` is just type assertion telling the TS compiler that it is of `Mock` type.
    // This would tell the TS compiler to use the mocked `useCart` instead of the OG `useCart'
    (useCart as Mock).mockReturnValue({ cartItems: [] });  
    const { container } = render(<CartDropdown onClose={vi.fn()} />);
    expect(container).toMatchSnapshot();
  });

  it("renders correctly with items in the cart", () => {
    (useCart as Mock).mockReturnValue({
      cartItems: [
        {
          id: 1,
          name: "Nike Air Max",
          brand: "Nike",
          size: "10",
          quantity: 2,
          price: 150,
          image: "/nike-air-max.jpg",
        },
      ]
    });

    const { container } = render(<CartDropdown onClose={vi.fn()} />);
    expect(container).toMatchSnapshot();
  });

  it("calls onClose when close button is clicked", async () => {
    const user = userEvent.setup();
    const onCloseMock = vi.fn();

    (useCart as Mock).mockReturnValue({ cartItems: [] });
    render(<CartDropdown onClose={onCloseMock} />);

    await user.click(screen.getByLabelText("Close cart"));
    expect(onCloseMock).toHaveBeenCalledTimes(1);
  });

  it("calls onClose when Continue Shopping button is clicked (empty cart)", async () => {
    const user = userEvent.setup();
    const onCloseMock = vi.fn();

    (useCart as Mock).mockReturnValue({ cartItems: [] });
    render(<CartDropdown onClose={onCloseMock} />);

    await user.click(screen.getByText(/Continue Shopping/i));
    expect(onCloseMock).toHaveBeenCalledTimes(1);
  });

  it("calls onClose when Continue Shopping button is clicked (cart with items)", async () => {
    const user = userEvent.setup();
    const onCloseMock = vi.fn();

    (useCart as Mock).mockReturnValue({
      cartItems: [
        {
          id: 1,
          name: "Nike Air Max",
          brand: "Nike",
          size: "10",
          quantity: 2,
          price: 150,
          image: "/nike-air-max.jpg",
        },
      ],
    });
    render(<CartDropdown onClose={onCloseMock} />);

    await user.click(screen.getByText(/Continue Shopping/i));
    expect(onCloseMock).toHaveBeenCalledTimes(1);
  });

  it("calls removeFromCart when remove button is clicked", async () => {
    const user = userEvent.setup();
    const removeFromCartMock = vi.fn();

    (useCart as Mock).mockReturnValue({
      cartItems: [
        {
          id: 1,
          name: "Nike Air Max",
          brand: "Nike",
          size: "10",
          quantity: 2,
          price: 150,
          image: "/nike-air-max.jpg",
        },
      ],
      removeFromCart: removeFromCartMock,
    });
    render(<CartDropdown onClose={vi.fn()} />);

    await user.click(screen.getByText(/Remove/i));
    expect(removeFromCartMock).toHaveBeenCalledWith(1, "10");
  });

  it("calls increaseQuantity when plus button is clicked", async () => {
    const user = userEvent.setup();
    const increaseQuantityMock = vi.fn();

    (useCart as Mock).mockReturnValue({
      cartItems: [
        {
          id: 1,
          name: "Nike Air Max",
          brand: "Nike",
          size: "10",
          quantity: 2,
          price: 150,
          image: "/nike-air-max.jpg",
        },
      ],
      increaseQuantity: increaseQuantityMock,
    });
    render(<CartDropdown onClose={vi.fn()} />);

    await user.click(screen.getByLabelText("Increase quantity"));
    expect(increaseQuantityMock).toHaveBeenCalledWith(1, "10");
  });

  it("calls decreaseQuantity when minus button is clicked", async () => {
    const user = userEvent.setup();
    const decreaseQuantityMock = vi.fn();

    (useCart as Mock).mockReturnValue({
      cartItems: [
        {
          id: 1,
          name: "Nike Air Max",
          brand: "Nike",
          size: "10",
          quantity: 2,
          price: 150,
          image: "/nike-air-max.jpg",
        },
      ],
      decreaseQuantity: decreaseQuantityMock,
    });
    render(<CartDropdown onClose={vi.fn()} />);

    await user.click(screen.getByLabelText("Decrease quantity"));
    expect(decreaseQuantityMock).toHaveBeenCalledWith(1, "10");
  });

  it("calls clearCart when Clear Cart button is clicked", async () => {
    const user = userEvent.setup();
    const clearCartMock = vi.fn();

    (useCart as Mock).mockReturnValue({
      cartItems: [
        {
          id: 1,
          name: "Nike Air Max",
          brand: "Nike",
          size: "10",
          quantity: 2,
          price: 150,
          image: "/nike-air-max.jpg",
        },
      ],
      clearCart: clearCartMock,
    });
    render(<CartDropdown onClose={vi.fn()} />);

    await user.click(screen.getByText(/Clear Cart/i));
    expect(clearCartMock).toHaveBeenCalledTimes(1);
  });
});