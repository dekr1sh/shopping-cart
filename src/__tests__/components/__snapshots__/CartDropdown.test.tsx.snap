// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CartDropdown > renders correctly when cart is empty 1`] = `
<div>
  <div
    class="absolute right-0 mt-2 w-80 sm:w-96 bg-background border rounded-lg shadow-lg z-50"
  >
    <div
      class="p-4 max-h-[80vh] overflow-auto"
    >
      <div
        class="flex justify-between items-center mb-4"
      >
        <h3
          class="font-bold text-lg"
        >
          Your Cart
        </h3>
        <button
          aria-label="Close cart"
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 size-9 cursor-pointer"
          data-slot="button"
        >
          <svg
            class="lucide lucide-x"
            fill="none"
            height="18"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M18 6 6 18"
            />
            <path
              d="m6 6 12 12"
            />
          </svg>
        </button>
      </div>
      <div
        class="py-8 text-center"
      >
        <svg
          class="lucide lucide-shopping-bag mx-auto mb-4 text-muted-foreground"
          fill="none"
          height="48"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="48"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6 2 3 6v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V6l-3-4Z"
          />
          <path
            d="M3 6h18"
          />
          <path
            d="M16 10a4 4 0 0 1-8 0"
          />
        </svg>
        <p
          class="text-muted-foreground"
        >
          Your cart is empty
        </p>
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs h-9 px-4 py-2 has-[>svg]:px-3 mt-4 hover:bg-primary/85 cursor-pointer"
          data-slot="button"
        >
          Continue Shopping
        </button>
      </div>
    </div>
  </div>
     
</div>
`;

exports[`CartDropdown > renders correctly with items in the cart 1`] = `
<div>
  <div
    class="absolute right-0 mt-2 w-80 sm:w-96 bg-background border rounded-lg shadow-lg z-50"
  >
    <div
      class="p-4 max-h-[80vh] overflow-auto"
    >
      <div
        class="flex justify-between items-center mb-4"
      >
        <h3
          class="font-bold text-lg"
        >
          Your Cart
        </h3>
        <button
          aria-label="Close cart"
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 size-9 cursor-pointer"
          data-slot="button"
        >
          <svg
            class="lucide lucide-x"
            fill="none"
            height="18"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M18 6 6 18"
            />
            <path
              d="m6 6 12 12"
            />
          </svg>
        </button>
      </div>
      <div
        class="space-y-4 mb-4"
      >
        <div
          class="flex items-center gap-4 border-b pb-4"
        >
          <div
            class="relative h-16 w-16 bg-muted rounded"
          >
            <img
              alt="Nike Air Max"
              class="object-cover rounded h-full w-full"
              src="/nike-air-max.jpg"
            />
          </div>
          <div
            class="flex-1"
          >
            <h4
              class="font-medium"
            >
              Nike Air Max
            </h4>
            <p
              class="text-sm text-muted-foreground"
            >
              Nike
               • Size: 
              10
            </p>
            <div
              class="flex justify-between items-center mt-2"
            >
              <div
                class="flex items-center border rounded"
              >
                <button
                  aria-label="Decrease quantity"
                  class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 size-9 h-8 w-8 rounded-none cursor-pointer"
                  data-slot="button"
                >
                  <svg
                    class="lucide lucide-minus"
                    fill="none"
                    height="14"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="14"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M5 12h14"
                    />
                  </svg>
                </button>
                <span
                  class="w-8 text-center"
                >
                  2
                </span>
                <button
                  aria-label="Increase quantity"
                  class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 size-9 h-8 w-8 rounded-none cursor-pointer"
                  data-slot="button"
                >
                  <svg
                    class="lucide lucide-plus"
                    fill="none"
                    height="14"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="14"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M5 12h14"
                    />
                    <path
                      d="M12 5v14"
                    />
                  </svg>
                </button>
              </div>
              <div
                class="text-right"
              >
                <p
                  class="font-medium"
                >
                  ₹
                  300.00
                </p>
                <button
                  class="text-xs text-muted-foreground hover:text-destructive cursor-pointer"
                >
                  Remove
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="border-t pt-4"
      >
        <div
          class="flex justify-between mb-2"
        >
          <span>
            Subtotal
          </span>
          <span
            class="font-medium"
          >
            ₹
            300.00
          </span>
        </div>
        <div
          class="flex justify-between mb-4"
        >
          <span>
            Shipping
          </span>
          <span>
            Calculated at checkout
          </span>
        </div>
        <div
          class="flex justify-between text-lg font-bold mb-4"
        >
          <span>
            Total
          </span>
          <span>
            ₹
            300.00
          </span>
        </div>
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3 w-full mb-2 cursor-pointer"
          data-slot="button"
        >
          Checkout
        </button>
        <div
          class="flex gap-2 mb-2"
        >
          <button
            class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 h-9 px-4 py-2 has-[>svg]:px-3 flex-1 cursor-pointer"
            data-slot="button"
          >
            Continue Shopping
          </button>
          <button
            class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-destructive text-white shadow-xs hover:bg-destructive/90 focus-visible:ring-destructive/20 dark:focus-visible:ring-destructive/40 dark:bg-destructive/60 h-9 px-4 py-2 has-[>svg]:px-3 flex-1 cursor-pointer"
            data-slot="button"
          >
            Clear Cart
          </button>
        </div>
      </div>
    </div>
  </div>
     
</div>
`;
